### DISCUSSION TITLE
How to reliably detect the state "after install and reboot, but before mark good"

### DISCUSSION THREAD

**Discussion by martin-hollingsworth-snke-obsolete** [rauc/rauc#1199] [Q&A] [answered]
Hello community,
as the last step of the update I usually call these lines to permanently finalize the A/B switch: (similar to the suggestion from meta-rauc at https://github.com/rauc/meta-rauc/blob/master/recipes-core/rauc/files/rauc-mark-good.service):
`root@device:~# rauc status mark-good booted`
`rauc status: marked slot rootfs.1 as good`
`root@device:~# rauc status mark-bad other`
`rauc status: marked slot rootfs.0 as bad`

However, from a script I want to catch the moment after install and reboot, but before mark-good and do some extra work on some sub-device (see https://rauc.readthedocs.io/en/latest/advanced.html#updating-sub-devices).
**What is the reliably way of doing so from a bash script?** (I am running rauc v.1.7).

I could catch the state where "boot status: good" is selected on two slots - any better solution?
`root@device:~# rauc status --output-format=shell --detailed | grep RAUC_SLOT_BOOT_STATUS`
`RAUC_SLOT_BOOT_STATUS_1='good'`
`RAUC_SLOT_BOOT_STATUS_2=''`
`RAUC_SLOT_BOOT_STATUS_3='good'`

**Comment by ejoerns** [comment, marked as answer]
@Hollingsworth-martin-bl If I got it right, you do not only require to run before the mark-good but also need to reliably know if you run the system the first time after an installation has been completed, right?
Or would it also be sufficient to always run the script before a mark-good (thus on each system start)?

If using systemd, you could probably leverage `system-update.target` from [systemd.offline-updates](https://www.freedesktop.org/software/systemd/man/systemd.offline-updates.html) to run something on the first boot only. But that might work for writable file systems only.

Relying on the boot status for detecting the life cycle state of an installation might work in some cases but is probably not too stable and depends on the system and bootloader configuration.

With RAUC v1.10 we have introduced a 'Transaction UUID' with exactly the purpose of being able to better track the life-cycle of an installation. For now, it is only generated and logged but the idea is to use it later e.g. for knowing if an installation is pending, which exact installation we marked good, etc. Maybe your use case would also fit in this.

**Reply by martin-hollingsworth-snke-obsolete** [reply]
@ejoerns thanks for the details! It gives me some ideas.

To answer your question:
> but also need to reliably know if you run the system the first time after an installation has been completed, right?

Exactly this! I am running some integrity checks and in case they fail I need to decide what to do. Fall back to the other slot (i.e. back to the old version) is only an option in above quoted situation - because outside of this situation I cannot rely on what is installed inside the other slot.

On a side question, why does the rauc-mark-good.service suggested in meta-rauc only run `rauc status mark-good`, but not mark-bad on the other slot? Thus keeping the other slot as active fallback during regular device operation? I.e. in my use case the system may never switch A/B without a service technician present.

**Reply by ejoerns** [reply]
> On a side question, why does the rauc-mark-good.service suggested in meta-rauc only run rauc status mark-good, but not mark-bad on the other slot? Thus keeping the other slot as active fallback during regular device operation? I.e. in my use case the system may never switch A/B without a service technician present.

Yes, this definitely is a valid and probably also a common use case, but when already having two redundant copies, other use cases might prefer keeping the inactive partition as a fallback also during regular device operation.

I would tend to say that, especially when it comes to security, having an always active fallback can be more and more problematic. So one solution for this case would be to always keep both partitions in sync. Then you can benefit from fallback (e.g. when fs gets corrupted for whatever reason) but ensure you do not fall back to an old version.

In general, I would also like to see the behavior that you described being somehow configurable in RAUC. Maybe already as part of the `mark-good` action or something similar. This might play well together with the things I noted before since you can then also have something like a "mark this installation transaction as good" in RAUC.

Note that the mark-good service provided with RAUC can (and should) be replaced with a custom one if the use case requires this.
It might also need to be entirely removed if e.g. the application should execute that mark-good instead.

**Reply by martin-hollingsworth-snke-obsolete** [reply]
> Note that the mark-good service provided with RAUC can (and should) be replaced with a custom one if the use case requires this.
It might also need to be entirely removed if e.g. the application should execute that mark-good instead.

Thanks, that confirms my impression.
